### ISSUE TITLE
Custom entries make PG and CoqIde hang?

### ISSUE THREAD

**Issue by bcpierce00** [coq/coq#10906]
The last two lines in the file below cause Proof General and CoqIde to hang, although plain coqc from the command line handles it just fine.

#### Coq Version

8.9.1.  I also confirmed the PG issue with 8.10.0.

_______________________________________________________________

Require Import String.

Inductive ty : Type :=
  | Bool  : ty
  | Arrow : ty -> ty -> ty.

Inductive tm : Type :=
  | var : string -> tm
  | app : tm -> tm -> tm
  | abs : string -> ty -> tm -> tm
  | tru : tm
  | fls : tm
  | test : tm -> tm -> tm -> tm.

Definition x : string := "x".
Definition y : string := "y".
Definition z : string := "z".

(* This might not be needed, given the hack below *)
Coercion var : string >-> tm.

Declare Custom Entry stlc.
Notation "[[ e ]]" := e (e custom stlc at level 10).
Notation "( x )" := x (in custom stlc, x at level 10).
Notation "x" := x (in custom stlc at level 0, x constr at level 0).

Notation "S -> T" :=
  (Arrow S T) (in custom stlc at level 2, right associativity).

Check [[ Bool -> Bool ]].

Notation "x y" := (app x y) (in custom stlc at level 1, left associativity).

Definition X : string := "X".
Definition Y : string := "Y".

Check [[ X ]].
Check [[ X Y ]].
Check [[ (X Y) (X Y) ]].

Notation "\ x : t . y" := (abs x t y) (in custom stlc at level 1, x global, left associativity).

Check [[ \X:Bool. X ]].
Check [[ (\X:Bool. X) Y ]].

**Comment by gares** [comment]
UIs chop the text by using a blind regexp, they look for a dot followed by a space. Coqc uses the actual parser to determine where a sentence ends.

On this front the LSP protocol solves the issue and we are working on adopting it.

Wrt this issue I believe it is not custom entry specific. The only workaround I see is to use something other than . in the notation for lambda.

**Comment by bcpierce00** [comment]
Aha — of course!  Many thanks for the quick reply, Enrico.

The issue with period in notations is probably documented someplace, but perhaps it deserves an extra mention in the custom entries section, to save others the trouble of being reminded of this restriction the hard way…?

**Comment by maximedenes** [comment]
Yes, we are working on separating parsing and execution, which would make it possible for IDEs to identify sentences reliable without executing them.
